[AIEX] Prevent memops and fmuladd jl inside ZOL and other enhancements

andcarminati opened the pull request:
This fixes the generation of some calls inside ZOL loops.

Related C test:
```
struct foo {
    long int buffer[1024];
};

void copy(foo *a, foo *b, int n) {
    for (int i = 0; i < n; i++) {
        *a = *b;
    }
}
```

Update 1: AIE2P support for JNZD was added for completeness.

Update 2: fmuladd is now tracked as a source of JL (see issue https://github.com/Xilinx/llvm-aie/issues/475).

andcarminati committed:
[AIEX] Prevent memops and fmuladd jl inside ZOL

F-Stuckmann commented on llvm/test/CodeGen/AIE/aie2/hardware-loops/loop-with-call.ll:
Since this now also includes aie2p, could we move it to `llvm/test/CodeGen/AIE/hardware-loops`

andcarminati replied:
Oh. indeed.

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseTargetTransformInfo.cpp:
```diff
@@ -64,6 +65,13 @@ static cl::opt<unsigned> PreferSwpOverUnroll(
 
 bool AIETTICommon::isLoweredToCall(const Function *F) {
+
+  if (F->getIntrinsicID() == Intrinsic::memcpy ||
```
talked offline: we may want to revert hardware loops after instruction selection, since at that point we know, if there will be an illegal call in the hardware loops.

niwinanto commented on llvm/lib/Target/AIE/AIEBaseTargetTransformInfo.cpp:
```diff
@@ -64,6 +65,13 @@ static cl::opt<unsigned> PreferSwpOverUnroll(
+    return true;
+  }
+
   return !F->isIntrinsic();
```
I am wondering why this check is not enough?

andcarminati replied:
Memcpy/set/move is an intrinsic that is lowered to call.

andcarminati committed:
[AIEX] Refactor target-agnostic ZOL tests to a common place

andcarminati committed:
[AIEX][AIE2P] Add JNZD HL support

This is intended to make HL support complete and refactored as much as possible.

andcarminati commented on llvm/lib/Target/AIE/AIEBaseInstructionSelector.cpp:
```diff
@@ -170,8 +170,94 @@ bool AIEBaseInstructionSelector::selectBrCondLoopDecrement(
+      [](const MachineInstr &MI,
+         const MachineRegisterInfo &MRI) -> std::optional<JNZDArgs> {
+    if (MI.getOpcode() != TargetOpcode::G_ICMP)
+      return {};
```
nullopt?

martien-de-jong replied:
Only if you also expand the auto

F-Stuckmann commented on llvm/lib/Target/AIE/AIE2InstrInfo.cpp:
```diff
@@ -1434,6 +1427,19 @@ AIE2InstrInfo::getZOLSupport() const {
 }
 
+std::optional<AIEBaseInstrInfo::JNZDSupport>
+AIE2InstrInfo::getJNZDSupport() const {
```
I love the JNZSupport Class

andcarminati replied:
Yes! @martien-de-jong' s idea.

martien-de-jong replied:
The term 'aspect oriented programming' comes to mind.

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseInstructionSelector.cpp:
```diff
@@ -170,8 +170,94 @@ bool AIEBaseInstructionSelector::selectBrCondLoopDecrement(
+
+  auto JNZDSupport = TII.getJNZDSupport();
+  if (!JNZDSupport) {
+    return false;
```
nit: add a debug message to notify that JNZD support is not supported

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseInstructionSelector.cpp:
```diff
@@ -804,3 +890,34 @@ void AIEBaseInstructionSelector::insertPtrAddForOffset(MachineRegisterInfo &MRI,
+  }
+
+  // Not a constant trip count, decrement at runtime
+  auto ADDI = MIB.buildInstr(JNZDSupport->DecTripCount, {I.getOperand(0)},
```
could this ADD lead to underflows?

andcarminati replied:
Basically no. Hardware loops are always rotated, so in the worst-case (1 iteration), the underflow will happen in the JNZD instruction, but after this point we don't care about the result.

F-Stuckmann commented on llvm/test/CodeGen/AIE/hardware-loops/zol-loop.ll:
```diff
@@ -0,0 +1,173 @@
+;
+; (c) Copyright 2024-2025 Advanced Micro Devices, Inc. or its affiliates
+
+; RUN: llc -O2 -mtriple=aie2 --enable-aie-hardware-loops \
```
can we work with common prefixes?

andcarminati replied:
This triggers some update problem:
```
WARNING: Prefix COMMON had conflicting output from different RUN lines for all functions in test /scratch/llvm-aie/llvm/test/CodeGen/AIE/hardware-loops/zol-loop.ll
```

Curious....

F-Stuckmann commented on llvm/test/CodeGen/AIE/hardware-loops/unknown-tc.ll:
```diff
@@ -0,0 +1,106 @@
+; (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
+
+; RUN: llc -O2 -mtriple=aie2 --issue-limit=1 %s -o - | FileCheck %s --check-prefix=AIE2
+; RUN: llc -O2 -mtriple=aie2p --issue-limit=1 %s -o - | FileCheck %s --check-prefix=AIE2
```
nit: update the prefix. It now implies, that there are no AIE2P checks

F-Stuckmann commented on llvm/test/CodeGen/AIE/hardware-loops/zol-md-accept-reject.ll:
```diff
@@ -2,9 +2,11 @@
+; RUN:    --enable-aie-zero-overhead-loops %s -o - | FileCheck %s
+; RUN: llc -O2 -stop-after=hardware-loops -mtriple=aie2p \
 ; RUN:    --enable-aie-zero-overhead-loops %s -o - | FileCheck %s
 
```
nit: could you provide a comment what this file is supposed to check?

F-Stuckmann commented on llvm/test/CodeGen/AIE/hardware-loops/unknown-tc.ll:
```diff
@@ -0,0 +1,106 @@
+; See https://llvm.org/LICENSE.txt for license information.
+; SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+;
+; (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: update date

F-Stuckmann commented on llvm/test/CodeGen/AIE/hardware-loops/simple.ll:
```diff
@@ -0,0 +1,94 @@
+; See https://llvm.org/LICENSE.txt for license information.
+; SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+;
+; (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: 2025

F-Stuckmann commented on llvm/test/CodeGen/AIE/hardware-loops/loop-with-call.ll:
```diff
@@ -0,0 +1,381 @@
+; AIE2-NEXT:    mova r1, #56; mov p2, p0
+; AIE2-NEXT:    st r17, [sp, #-24]; jl #strcpy // 4-byte Folded Spill
+; AIE2-NEXT:    st p6, [sp, #-28]; mul r1, r17, r1 // 4-byte Folded Spill Delay Slot 5
+; AIE2-NEXT:    st lr, [sp, #-16]; movxm p6, #symbolTable // 4-byte Folded Spill Delay Slot 4
```
Is there a more compact way of showing that these functions reject hardware loops?

F-Stuckmann commented on llvm/test/CodeGen/AIE/hardware-loops/instruction-select-zol-start.mir:
```diff
@@ -0,0 +1,46 @@
+# (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
+
+# RUN: llc -O2 -mtriple=aie2 -run-pass=instruction-select %s -verify-machineinstrs -o - | FileCheck %s --check-prefix=AIE2
+# RUN: llc -O2 -mtriple=aie2p -run-pass=instruction-select %s -verify-machineinstrs -o - | FileCheck %s --check-prefix=AIE2p
```
nit: use common for the prefix

andcarminati commented on llvm/test/CodeGen/AIE/hardware-loops/sibling.ll:
```diff
@@ -0,0 +1,152 @@
+; See https://llvm.org/LICENSE.txt for license information.
+; SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+;
+; (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: date.

andcarminati commented on llvm/test/CodeGen/AIE/hardware-loops/nested.ll:
```diff
@@ -0,0 +1,165 @@
+; See https://llvm.org/LICENSE.txt for license information.
+; SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+;
+; (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: date.

andcarminati commented on llvm/test/CodeGen/AIE/hardware-loops/instruction-select-zol-start.mir:
```diff
@@ -0,0 +1,46 @@
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: date.

martien-de-jong commented on llvm/lib/Target/AIE/AIEBaseInstructionSelector.cpp:
```diff
@@ -170,8 +170,94 @@ bool AIEBaseInstructionSelector::selectBrCondLoopDecrement(
+  }
+
+  Register CondReg = BrCond.getOperand(0).getReg();
+  MachineBasicBlock *DestMBB = BrCond.getOperand(1).getMBB();
```
I wouldn't mind moving this closer to its first use.
nit: const CondReg and * const DestMBB

martien-de-jong commented on llvm/lib/Target/AIE/aie2p/AIE2PInstrInfo.cpp:
```diff
@@ -1743,6 +1735,19 @@ AIE2PInstrInfo::getZOLSupport() const {
+  Result.MovBlockAddrOpcode = AIE2P::MOVXM;
+  Result.PointerRegisterClass = &AIE2P::eP_as_32BitRegClass;
+  Result.LoopDecOpcode = AIE2P::LoopDec;
+  Result.LoopJNZOpcode = AIE2P::LoopJNZ;
```
Why is this AIE2P::LoopJNZ? Shouldn't it just be  JNZ?

andcarminati replied:
We expand very late all pseudo jumps/branches because we need to generate `DelayedSchedBarrier` after. This pseudo is combined later on to generate PseudoJNZD that will expand to  `JNZD, DelayedSchedBarrier`.

martien-de-jong commented on llvm/lib/Target/AIE/AIEBaseInstructionSelector.cpp:
```diff
@@ -170,8 +171,94 @@ bool AIEBaseInstructionSelector::selectBrCondLoopDecrement(
+// Try to match BRCOND(Intrinsic::loop_decrement)
+bool AIEBaseInstructionSelector::selectBrCondLoopDecrementReg(
+    MachineInstr &BrCond, MachineRegisterInfo &MRI) {
+
```
assert  that it is a BRCOND?

martien-de-jong commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.cpp:
```diff
@@ -386,6 +386,16 @@ bool AIEBaseInstrInfo::isHardwareLoopEnd(unsigned Opcode) const {
+  return JNZDSupport && Opcode == JNZDSupport->LoopDecOpcode;
+}
+
+bool AIEBaseInstrInfo::isHardwareLoopJNZ(unsigned Opcode) const {
```
could we rename JNZ -> JNZD?

andcarminati replied:
This is really a JNZ, we have two different pseudos here.

martien-de-jong commented on llvm/lib/Target/AIE/AIEBaseInstructionSelector.cpp:
```diff
@@ -170,8 +171,94 @@ bool AIEBaseInstructionSelector::selectBrCondLoopDecrement(
+    JNZDArgs Args;
+    Args.NewLC = CmpLHS->getOperand(0).getReg();
+    Args.PrevLC = CmpLHS->getOperand(2).getReg();
+    Args.IntrinInst = CmpLHS;
```
Curious: it looks as if IntrinInst carries all necessary information. do you copy the relevant operands for clarity, for efficiency or for safety?

andcarminati replied:
Honestly, you are 100% correct! I don't have the answer because I just refactored the code. I guess your first idea is the best explanation: clarity!

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseHardwareLoops.cpp:
```diff
@@ -435,12 +435,14 @@ void AIEBaseHardwareLoops::expandLoopEnd(LowOverheadLoop &LoLoop) {
     return;
   }
+  auto LoweringData = TII->getJNZDSupport();
+  assert(LoweringData);
```
CHECK: this would crash under AIE1?

andcarminati replied:
AFAIK we don't implement HL for AIE1, so we are safe!

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.h:
```diff
@@ -359,6 +376,10 @@ struct AIEBaseInstrInfo : public TargetInstrInfo {
   // registers(lc, le, ls, etc.) and the end of the loop,
   virtual unsigned getLoopSetupDistance() const;
 
+  // All opcodes etc used for ZNZD lowering. If this returns none, we have no
```
nit: JNZD

andcarminati replied:
Eagle eyes!

martien-de-jong reviewed (dismissed):
LGTM

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.h:
```diff
@@ -60,6 +60,23 @@ struct AIEBaseInstrInfo : public TargetInstrInfo {
+    const TargetRegisterClass *PointerRegisterClass;
+    // Target Opcode for LoopDec.
+    unsigned LoopDecOpcode;
+    // Target Opcode for JNZ.
```
nit: JNZD

andcarminati replied:
If you see we have both opcodes here, JNZ and JNZD.

F-Stuckmann replied:
Ah I got confused with the Class name

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.h:
```diff
@@ -60,6 +60,23 @@ struct AIEBaseInstrInfo : public TargetInstrInfo {
+    // Target Opcode for JNZ.
+    unsigned LoopJNZOpcode;
+    // Decrement Trip Count Opcode.
+    unsigned DecTripCount;
```
nit: Shouldn't this also have Opcode in the name?

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.h:
```diff
@@ -60,6 +60,23 @@ struct AIEBaseInstrInfo : public TargetInstrInfo {
+    unsigned LoopDecOpcode;
+    // Target Opcode for JNZ.
+    unsigned LoopJNZOpcode;
+    // Decrement Trip Count Opcode.
```
nit: harmonize to `// Target Opcode for Decrement Trip Count `

F-Stuckmann commented on llvm/lib/Target/AIE/AIEBaseInstructionSelector.cpp:
```diff
@@ -170,8 +171,94 @@ bool AIEBaseInstructionSelector::selectBrCondLoopDecrement(
 bool AIEBaseInstructionSelector::selectG_BRCOND(MachineInstr &I,
                                                 MachineRegisterInfo &MRI) {
+  // Try matching JNZD loop end
+  if (selectBrCondLoopDecrementReg(I, MRI)) {
```
nit: since this is a JNZD check for the BranchCondition, shouldn't the function be called something like `isJNZDSelectable()`, then you can also remove the redundant comment.

Similarly  can we have a nicer function name for `selectBrCondLoopDecrement`?

andcarminati replied:
Usually we use the name that we are trying to select (Generic), not the result. In this case we select a Branch that succeeds a decrement by a LoopJNZ. Please note that We turn it into a PseudoJNZD way after selection.

F-Stuckmann reviewed (approved):
If more special cases happen, where we have to predict if instruction selection is implemented with a call, we might want to spend the effort in selecting JNZD after we can verify that we do not have a call in the loop.
